pkg/auth/iam/iam.go

```diff
@@ -90,57 +90,81 @@ func (filter *Filter) Auth(opts ...FilterOption) restful.FilterFunction {
 			return
 		}
 
-		claims, err := filter.iamClient.ValidateAndParseClaims(token)
-		if err != nil {
-			logrus.Warn("unauthorized access: ", err)
-			logIfErr(resp.WriteHeaderAndJson(http.StatusUnauthorized, ErrorResponse{
-				ErrorCode:    UnauthorizedAccess,
-				ErrorMessage: ErrorCodeMapping[UnauthorizedAccess],
-			}, restful.MIME_JSON))
+		filter.validateUser(opts, req, resp, chain, token, tokenFrom)
+	}
+}
 
+// PublicAuth returns a filter that allow unauthenticate request and request with valid access token in auth header or cookie
+// If request has acces token, the token's claims will be passed in the request.attributes["JWTClaims"] = *iam.JWTClaims{}
+// This filter is expandable through FilterOption parameter
+// Example:
+// iam.PublicAuth(
+// 		WithValidUser(),
+//		WithPermission("ADMIN"),
+// )
+func (filter *Filter) PublicAuth(opts ...FilterOption) restful.FilterFunction {
+	return func(req *restful.Request, resp *restful.Response, chain *restful.FilterChain) {
+		token, tokenFrom, err := parseAccessToken(req)
+		if err != nil {
+			chain.ProcessFilter(req, resp)
 			return
 		}
 
-		req.SetAttribute(ClaimsAttribute, claims)
+		filter.validateUser(opts, req, resp, chain, token, tokenFrom)
+	}
+}
 
-		if tokenFrom == tokenFromCookie {
-			valid := filter.validateRefererHeader(req, claims)
-			if !valid {
-				logIfErr(resp.WriteHeaderAndJson(http.StatusUnauthorized, ErrorResponse{
-					ErrorCode:    InvalidRefererHeader,
-					ErrorMessage: ErrorCodeMapping[InvalidRefererHeader],
-				}, restful.MIME_JSON))
+func (filter *Filter) validateUser(opts []FilterOption, req *restful.Request, resp *restful.Response, chain *restful.FilterChain, token, tokenFrom string) {
```

Comment:
can't find the usage of "tokenFrom", do we need it on this method?

Reply:
yes, `tokenFrom` is used for checking if token from cookies, it should be in line 131